exclude audio from shuffle

briaguya0 opened the pull request:
![image](https://user-images.githubusercontent.com/70942617/219153968-2c2b7869-1303-4698-8954-5733ebef046a.png)

todo:
- [x] optimize (need the menu for excluding to not be slow even with hundreds of sequences loaded)
- [x] prevent hangs (it's currently possible to exclude everything and turn the randomize button into a "start an infinite loop of looking for a valid sequence when there are none" button)
- [x] figure out what happens when loading a new otr of sequences after already having an exclude list (this needs to be handled properly before this can be considered merge ready, losing an exclude list after updating to get a couple new sequences from DJ would be terrible UX)
- [x] add preview buttons to entries on exclude tab (probably worth moving the exiting preview button functionality out into a method that can be used in both places)
- [x] add "chip" style type (sfx/bgm etc.) tags for each entry in the list
- [x] add filters for type (sfx/bgm etc.) 
- [x] use cvar-per-sequence/sfx for exclude list, save every time we update (this should prevent the laggy "make a giant string of excluded sequences" thing, which was the reason for not doing it every time in the first place. i think we use buffered file output so it shouldn't be an issue on PC (might be laggy on console))
- [x] pre-review cleanup
- [x] bulk include/exclude (ideally filter and then include/exclue all visible)
- [ ] get it reviewed/address feedback

future things (make issues for these when this gets merged):
- [ ] add presets that load from json so people can share their soundtracks (will check box once issue is made to do this)

<!--- section:artifacts:start -->
### Build Artifacts
  - [soh-linux-compatibility.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/559556361.zip)
  - [soh-linux-performance.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/559556366.zip)
  - [soh-mac.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/559556368.zip)
  - [soh-switch.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/559556370.zip)
  - [soh-wiiu.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/559556371.zip)
  - [soh-windows.zip](https://nightly.link/HarbourMasters/Shipwright/actions/artifacts/559556373.zip)
<!--- section:artifacts:end -->

Commit:
first pass at excluding sequences from shuffle

Commit:
fix mac build?

Commit:
use struct instead of tuple

Commit:
make it better

Commit:
remove commented out code

Commit:
move list init block so we get all the custom sequences

aMannus commented:
Is it possible to put the exclusions under their categories a-la excluded locations? Currently it's a bit hard to make out if something is a sound effect or a music track for example. 

So have all `SEQ_BGM_WORLD` entries under a dropdown like:
![image](https://user-images.githubusercontent.com/4244591/215427217-f1415c1c-12cf-493d-95d4-02fe0e21e231.png)

etc etc.

aMannus commented:
You also named the tab "Exclude Sequences From Shuffle" but in reality, you can also exclude sound effects in the tab, so the name isn't very accurate.

Maybe "Exclude SFX when randomizing" is more accurate (albeit a bit long)?

aMannus commented:
Maybe out of scope of this PR, but a button to remove all the exclusions at once could also be useful (and excluded locations in the Randomizer menu could also use this if we had it). This really is more of a suggestion though.

Commit:
preview buttons

Commit:
start adding disabled buttons for types

Commit:
make previews work for all the sequences in exclude tab

Commit:
better save logic

Commit:
add selectable table

Commit:
filter by type

Commit:
some names

briaguya0 commented:
@aMannus this is ready for another look

> Maybe "Exclude SFX when randomizing" is more accurate (albeit a bit long)?

I don't like that because i think of SFX as sound *effects*, and separate from music. I ended up renaming the whole window the "Audio Editor" and changed the name of the tab from "Exclude Sequences From Shuffle" to "Audio Shuffle Pool Management"

it's wordy and i'm not the biggest fan of the name but it at least feels accurate

>  a button to remove all the exclusions at once could also be useful

~i added that as a "future things" checkbox. i agree it'd be great but i also figure it can wait until after this is merged~ https://github.com/HarbourMasters/Shipwright/pull/2411/commits/2469dbcf0faa00b4e014bf41994263d2b81093f0

Commit:
include/exclude all buttons

Commit:
sort alphabetically

Commit:
initialize excludes list on window open instead of exclude tab open

Commit:
tried refactoring, broke things

Commit:
include the correct group name

Commit:
use audiocollection instead of removed var

Commit:
builds, seems to work

Commit:
bit of renaming

Commit:
imgui cleanup

Commit:
move some stuff from editor to collection

Commit:
fix sequence name stuff

dcvz commented on soh/soh/Enhancements/audio/AudioCollection.cpp:
```diff
@@ -0,0 +1,285 @@
+#include <vector>
+#include <Utils/StringHelper.h>
+#include <libultraship/bridge.h>
+#include <ImGuiImpl.h>
```
We should prefer `#include <libultraship/classes.h>` just because this is currently wrong way in which SoH is using LUS.. eventually we should only be importing things that are actually exposed in the `include` folder.

briaguya0 replied:
when i tried that i wasn't able to build, i got a ton of SDL include errors...

i can try again, but it was definitely broken for me when i tried

briaguya0 replied:
tried again, it's working now, updated

dcvz commented on soh/soh/Enhancements/audio/AudioCollection.cpp:
```diff
@@ -0,0 +1,285 @@
+    SeqType type = SEQ_BGM_CUSTOM;
+    std::string typeString = splitFileName[splitFileName.size() - 1];
+    std::locale loc;
+    for (int i = 0; i < typeString.length(); i++) {
```
```suggestion
    for (size_t i = 0; i < typeString.length(); i++) {
```

dcvz commented on soh/soh/Enhancements/audio/AudioCollection.cpp:
```diff
@@ -0,0 +1,266 @@
+
+void AudioCollection::AddToCollection(char* otrPath, uint16_t seqNum) {
+    std::vector<std::string> splitName = StringHelper::Split(std::string(otrPath), "/");
+    std::string fileName = splitName[splitName.size() - 1];
```
You can rely on `std::filesystem::path` here to give you this functionality -- instead of doing manual path splitting.
```c++
std::string fileName = std::filesystem::path(otrPath).filename();
```

dcvz commented on soh/soh/Enhancements/audio/AudioCollection.cpp:
```diff
@@ -0,0 +1,266 @@
+void AudioCollection::AddToCollection(char* otrPath, uint16_t seqNum) {
+    std::vector<std::string> splitName = StringHelper::Split(std::string(otrPath), "/");
+    std::string fileName = splitName[splitName.size() - 1];
+    std::vector<std::string> splitFileName = StringHelper::Split(fileName, "_");
```
This is a suggestion: you could just rely on boost here to move away from our own helpers
```c++
std::vector<std::string> splitFileName;
boost::split(splitFileName, fileName, [](char c) { return c == '_'; });
```

dcvz commented on soh/soh/Enhancements/audio/AudioCollection.cpp:
```diff
@@ -0,0 +1,266 @@
+    std::locale loc;
+    for (int i = 0; i < typeString.length(); i++) {
+        typeString[i] = std::tolower(typeString[i], loc);
+    }
```
You could replace all of this logic with just:
```c++
boost::algorithm::to_lower(typeString);
```

dcvz commented on soh/soh/Enhancements/audio/AudioCollection.cpp:
```diff
@@ -0,0 +1,266 @@
+    if (typeString == "fanfare") {
+        type = SEQ_FANFARE;
+    }
+    SequenceInfo info = {seqNum,
+                         sequenceName,
+                         StringHelper::Replace(StringHelper::Replace(sequenceName, " ", "_"), "~", "-"),
+                         type};
```
Some more boost helpers:
```c++
std::string sfxKey = boost::replace_all_copy(sequenceName, " ", "_");
boost::replace_all(sfxKey, "~", "-");

SequenceInfo info = { seqNum, sequenceName, sfxKey, type };
```

dcvz commented on soh/soh/Enhancements/audio/AudioCollection.cpp:
```diff
@@ -0,0 +1,266 @@
+    return static_cast<uint16_t>(replacementSeq);
+}
+
+uint16_t AudioCollection::GetReverseReplacementSequence(uint16_t seqId) {
```
I'd name this `GetOriginalSequenceId` I had to read the function to understand what this was trying to do.

briaguya0 replied:
yeah, naming was copied over from what we had before. I'm not sure if I like `GetOriginalSequenceId` though. Something like `GetSequenceBeingReplacedId` feels more clear to me, thoughts?

dcvz commented on soh/soh/Enhancements/audio/AudioCollection.cpp:
```diff
@@ -0,0 +1,266 @@
+}
+
+void AudioCollection::InitializeShufflePool() {
+    if (!shufflePoolInitialized) {
```
Style thing: Just do an early return instead of encasing the whole thing in the if:
```c++
if (shufflePoolInitialized) {
    return;
}
```

briaguya0 replied:
definitely just copied this block over from when it was part of a much bigger method, good call!

dcvz commented on soh/soh/Enhancements/audio/AudioCollection.cpp:
```diff
@@ -0,0 +1,285 @@
+    if (seqIt != sequenceMap.end()) {
+        return seqIt->second.label.c_str();
+    }
+    return NULL;
```
```suggestion
    return nullptr;
```

dcvz commented on soh/soh/Enhancements/audio/AudioEditor.cpp:
```diff
@@ -0,0 +1,566 @@
+
+        ImGui::TableNextRow();
+        ImGui::TableNextColumn();
+        ImGui::Text(seqData.label.c_str());
```
```suggestion
        ImGui::Text("%s", seqData.label.c_str());
```

dcvz commented on soh/soh/Enhancements/audio/AudioEditor.cpp:
```diff
@@ -0,0 +1,566 @@
+                        ImGui::SameLine();
+                        DrawTypeChip(seqInfo->category);
+                        ImGui::SameLine();
+                        ImGui::Text(seqInfo->label.c_str());
```
```suggestion
                        ImGui::Text("%s", seqInfo->label.c_str());
```

dcvz commented on soh/soh/Enhancements/audio/AudioEditor.cpp:
```diff
@@ -0,0 +1,566 @@
+                        ImGui::SameLine();
+                        DrawTypeChip(seqInfo->category);
+                        ImGui::SameLine();
+                        ImGui::Text(seqInfo->label.c_str());
```
```suggestion
                        ImGui::Text("%s", seqInfo->label.c_str());
```

dcvz reviewed (changes requested):
Overall changes look good! Definitely puts us in a place where we can better optimise audio editing down the line 👯 

Left small comments here and there on things we could improve.

briaguya0 committed:
Update soh/soh/Enhancements/audio/AudioCollection.cpp

Co-authored-by: David Chavez <david@dcvz.io>

briaguya0 committed:
Update soh/soh/Enhancements/audio/AudioCollection.cpp

Co-authored-by: David Chavez <david@dcvz.io>

briaguya0 committed:
Apply suggestions from code review

Co-authored-by: David Chavez <david@dcvz.io>

Commit:
use filesystem path to get filename from otr

Commit:
formatting

Commit:
looks like that works now

Commit:
rename func, remove unused confusingly named func

Commit:
fix windows filesystem path string issue

dcvz reviewed (approved):
LGTM 👍
